ra/ra.go

```diff
@@ -1423,19 +1396,57 @@ func (ra *RegistrationAuthorityImpl) NewOrder(ctx context.Context, req *rapb.New
 	if err != nil {
 		return nil, err
 	}
+	for _, name := range parsedCSR.DNSNames {
+		if err := ra.PA.WillingToIssue(core.AcmeIdentifier{Value: name, Type: core.IdentifierDNS}); err != nil {
+			return nil, err
+		}
+	}
+
+	now := ra.clk.Now().UnixNano()
+	existingAuthz, err := ra.SA.GetAuthorizations(ctx, &sapb.GetAuthorizationsRequest{
+		RegistrationID: req.RegistrationID,
+		Now:            &now,
+		Domains:        parsedCSR.DNSNames,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	gotAuthzFor := make(map[string]bool, len(parsedCSR.DNSNames))
+	for _, v := range existingAuthz.Authz {
+		gotAuthzFor[*v.Domain] = true
+		order.Authorizations = append(order.Authorizations, *v.Authz.Id)
+	}
 
-	// TODO(#2955): Replace this with the batched methods
+	if len(gotAuthzFor) < len(parsedCSR.DNSNames) {
+		if err := ra.checkPendingAuthorizationLimit(ctx, *req.RegistrationID); err != nil {
+			return nil, err
+		}
+	}
+
+	var newAuthz []*corepb.Authorization
 	for _, name := range parsedCSR.DNSNames {
-		authz, err := ra.NewAuthorization(ctx, core.Authorization{
-			Identifier: core.AcmeIdentifier{
-				Type:  core.IdentifierDNS,
-				Value: name,
-			},
-		}, *req.RegistrationID)
+		if _, present := gotAuthzFor[name]; present {
```

Comment:
Let's switch this to `if gotAuthzFor[name] {`, taking advantage of the nice properties of using a map of bools.

Reply:
Ping on this comment